walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,135 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        // Create full (mandatory + optional) attendee list
+        Collection<String> fullAttendees = new ArrayList<>();
+        fullAttendees.addAll(request.getAttendees());
+        fullAttendees.addAll(request.getOptionalAttendees());
+
+        MeetingRequest fullRequest = new MeetingRequest(fullAttendees, request.getDuration());
+        Collection<TimeRange> resultWithOptionals = queryHelper(events, fullRequest);
+
+        // If any TimeRanges are available with optionals include, return those,
+        // Otherwise exculde optionals
+        // Notable case: When there are no mandatory attendees, optionals are treated as mandatory and returned.
+        if (request.getAttendees().isEmpty() || !resultWithOptionals.isEmpty()) {
+            return resultWithOptionals;
+        } else {
+            return queryHelper(events, request);
+        }
+    }
+
+
+    public Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request) {
+
+        // Remove events that do not share in required attendees.
+        ArrayList<Event> eventList = new ArrayList<>(events);
+        AttendeeHelper attendees = new AttendeeHelper(request.getAttendees());
+        
+        Iterator<Event> eventIt = eventList.iterator();
+        while (eventIt.hasNext()) {
+            if (!attendees.eventHasAttendees(eventIt.next())) {
+                eventIt.remove();
+            }
+        }
+        ArrayList<TimeRange> eventTimeRanges = new ArrayList<>();
+        for (Event event : eventList) {
+            eventTimeRanges.add(event.getWhen());
+        }
+        
+        ArrayList<TimeRange> occupiedTimeRanges = sortCombineOverlappingTimeRanges(eventTimeRanges);
+        ArrayList<TimeRange> result = new ArrayList<>();
+        
+        int startTime = TimeRange.START_OF_DAY;
+        // For each event, check for time between current event start and end of previous event (initially 0)
+        // Time between end of final event and end of day must be checked outside of loop.
+        for (TimeRange occupiedTimeRange : occupiedTimeRanges) {
+            if (occupiedTimeRange.start() - startTime >= request.getDuration()) {
+                result.add(TimeRange.fromStartEnd(startTime, occupiedTimeRange.start(), false));
+            }
+            startTime = occupiedTimeRange.end();
+        }
+
+        // Check for availability between last event and end of day.
+        if ((TimeRange.END_OF_DAY + 1 - startTime) >= request.getDuration()) {
+            result.add(TimeRange.fromStartEnd(startTime, TimeRange.END_OF_DAY, true));
+        }
+
+        return result;
+    }
+
+    private ArrayList<TimeRange> sortCombineOverlappingTimeRanges(ArrayList<TimeRange> eventRanges) {
+        Collections.sort(eventRanges, TimeRange.ORDER_BY_START);
```

Comment:
You can move this after you check if `eventRanges` is empty (no reason to try sorting until after you know there's items to sort)